### PR TITLE
[SPARK-33299][SQL] Support schema in JSON format by `from_json`/`from_csv` Spark SQL and PySpark

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Move schema parsing from `from_json` in Scala API to the common function in `ExprUtils.evalTypeExpr` which is used in the `JsonToStructs` and `CsvToStructs` expressions, and as a consequence in SQL API, for instance.

### Why are the changes needed?
Currently, `from_json` has different behavior in Scala API and in PySpark/SQL. In Scala API, it accepts schema in JSON format. To improve user experience with Spark SQL and PySpark, this PR proposes to unify the behavior of `from_json` and `from_csv`.

### Does this PR introduce _any_ user-facing change?
Yes.

Before (in Spark 3.0)
```sql
spark-sql> select from_json('{"a":1}', '{"type" : "map", "keyType" : "string", "valueType" : "integer", "valueContainsNull" : true}');
Error in query:
mismatched input '{' expecting {'ADD', 'AFTER', ...}(line 1, pos 0)

== SQL ==
{"type" : "map", "keyType" : "string", "valueType" : "integer", "valueContainsNull" : true}
^^^
```

After:
```sql
spark-sql> select from_json('{"a":1}', '{"type" : "map", "keyType" : "string", "valueType" : "integer", "valueContainsNull" : true}');
{"a":1}
```

### How was this patch tested?
Added test cases to `csv-functions.sql` and `json-functions.sq`

### CODE REVIEW

**Comment by MaxGekk** [comment]
@HyukjinKwon @maropu Please, review this PR.

**Comment by zero323** [comment]
We should probably update SparkR (in [collection functions doc](https://github.com/apache/spark/blob/master/R/pkg/R/functions.R#L220-L227)) and PySpark docs.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -4100,13 +4099,8 @@ object functions {
...
    */
-  def from_json(e: Column, schema: String, options: Map[String, String]): Column = {
-    val dataType = parseTypeWithFallback(
-      schema,
-      DataType.fromJson,
-      "Cannot parse the schema in JSON format: ",
-      fallbackParser = DataType.fromDDL)
-    from_json(e, dataType, options)
+  def from_json(e: Column, schema: String, options: Map[String, String]): Column = withExpr {
+    new JsonToStructs(e.expr, lit(schema).expr, options)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4103]: Could you update `ExpressionDescription` (usage and examples) of `CsvToStructs` and `JsonToStructs`?

**Comment by MaxGekk** [comment]
@HyukjinKwon already updated python docs :-) https://github.com/apache/spark/pull/18498/files#diff-af2c27f925c4634d7efc25df50dc66b99e5a79f6f897dc4c49302492a48ca3f6R1889

**Comment by MaxGekk** [comment]
I guess SparkR especially handles schema strings. It seems it doesn't support schemas in JSON format. @HyukjinKwon , correct? https://github.com/apache/spark/blob/3beab8d8a8e2ed5e46e063d5a44face40c5fac90/R/pkg/R/functions.R#L2522

I am going to revert my changes in SparkR.

**Comment by zero323** [comment]
> @HyukjinKwon already updated python docs :-)

I was thinking more about example of using JSON input for schema like the one you've added in 
https://github.com/apache/spark/pull/30201/commits/5820120e3f6fb4966f886e26f59ba4aa2c8d8625.

But it doesn't seem like there's any user-visible change in Python behavior here after all, is there?

> I guess SparkR especially handles schema strings. 

Yes, my bad.

**Comment by HyukjinKwon** [comment]
Thanks @MaxGekk for catching this. Will take a look tomorrow.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4103]: Wait .. I remember actually I had a discussion about this (with @maropu ..?) before. JSON format schema is actually pretty internal ser/de purpose, and just encouraging DDL format is better, and JSON format here was kept just for the compatibility reason.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4103]: Okay but `from_json` itself supports anyway. Okay should be best to match across other languages for the same API.

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -2860,6 +2860,10 @@ def from_json(col, schema, options={}):
     [Row(json=Row(a=1))]
     >>> df.select(from_json(df.value, "MAP<STRING,INT>").alias("json")).collect()
     [Row(json={'a': 1})]
+    >>> df.select(from_json(df.value,\
+     '{"type" : "map", "keyType" : "string", "valueType" : "integer", "valueContainsNull" : true}')\
+     .alias("json")).collect()
+    [Row(json={'a': 1})]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:2866]:
```suggestion
    >>> schema = (
    ...     '{"type" : "map", "keyType" : "string",'
    ...     '"valueType" : "integer", "valueContainsNull" : true}')
    >>> df.select(from_json(df.value, schema).alias("json")).collect()
    [Row(json={'a': 1})]
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 3]
```diff
@@ -505,13 +505,15 @@ case class JsonTuple(children: Seq[Expression])
  */
 // scalastyle:off line.size.limit
 @ExpressionDescription(
-  usage = "_FUNC_(jsonStr, schema[, options]) - Returns a struct value with the given `jsonStr` and `schema`.",
+  usage = "_FUNC_(jsonStr, schema[, options]) - Returns a struct value with the given `jsonStr` and `schema`. The schema can be in the table schema format, data type or in the JSON format.",
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:508]: Let's remove these docs and examples. I think it's better to don't encourage JSON sed/de.

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:2866]:
```suggestion
    >>> schema =  (
    ...     '{"type" : "map", "keyType" : "string",'
    ...     '"valueType" : "integer", "valueContainsNull" : true}')
    >>> df.select(from_json(df.value, schema).alias("json")).collect()
    [Row(json={'a': 1})]
```

**Comment by HyukjinKwon** [comment]
Okay, I have thought a while + had a offline discussion with @MaxGekk. Looks like we should better just fix the doc in PySpark, and don't support JSON format in PySpark, SQL, and SparkR for now. We can keep it in Scala side for a legacy reason.

The reason is that, I do think JSON format is sort of internal, and we should better promote DDL string types instead. I haven't heard any complaint about JSON string support so far (surprisingly).

**Comment by MaxGekk** [comment]
Please, review this https://github.com/apache/spark/pull/30226

**Pull request closed without merging**